src/core/node_impl.h

```diff
@@ -0,0 +1,135 @@
+#pragma once
+
+#include "global_include.h"
+#include "mavlink_address.h"
+#include "mavlink_include.h"
+#include "mavlink_parameters.h"
+#include "mavlink_commands.h"
+#include "mavlink_message_handler.h"
+#include "mavlink_mission_transfer.h"
+#include "timeout_handler.h"
+#include "call_every_handler.h"
+#include "thread_pool.h"
+#include "timesync.h"
+#include "system.h"
+#include <cstdint>
+#include <functional>
+#include <atomic>
+#include <vector>
+#include <unordered_set>
+#include <map>
+#include <thread>
+#include <mutex>
+#include <future>
+
+namespace mavsdk {
+
+class MavsdkImpl;
+class PluginImplBase;
+
+// This class is the impl of Node. this is to hide the private methods
+// and functionality from the public library API.
+class NodeImpl : public Sender {
+public:
+    explicit NodeImpl(
+            MavsdkImpl& parent, uint8_t system_id, uint8_t component_id);
+    ~NodeImpl();
+
+    void process_mavlink_message(mavlink_message_t& message);
+
+    typedef std::function<void(const mavlink_message_t)> mavlink_message_handler_t;
+
+    void register_mavlink_message_handler(
+        uint16_t msg_id, mavlink_message_handler_t callback, const void* cookie);
+
+    void unregister_mavlink_message_handler(uint16_t msg_id, const void* cookie);
+    void unregister_all_mavlink_message_handlers(const void* cookie);
+
+    void register_timeout_handler(std::function<void()> callback, double duration_s, void** cookie);
+    void refresh_timeout_handler(const void* cookie);
+    void unregister_timeout_handler(const void* cookie);
+
+    void add_call_every(std::function<void()> callback, float interval_s, void** cookie);
+    void change_call_every(float interval_s, const void* cookie);
+    void reset_call_every(const void* cookie);
+    void remove_call_every(const void* cookie);
```

Comment:
I'm wondering if all these API calls should be exposed by each node or where they fit it in :thinking:.